File: ABC/246/d.cpp
#include <bits/stdc++.h>
using namespace std;
using ll = long long;
using vll = vector<ll>;
using vvll = vector<vll>;

#ifdef _DEBUG
#define debug(x) cout<<#x<<": "<<x<<endl
#define debug_vec(x) cout<<#x<<" ->"<<endl;for(size_t i=0;i<x.size();i++){cout<<x.at(i)<<"; ";}cout<<endl;
#define debug_vecvec(x) cout<<#x<<" ->"<<endl;for(size_t i=0;i<x.size();i++){for(size_t j=0;j<x[i].size();j++){cout<<x.at(i).at(j)<<"; ";}cout<<endl;}cout<<endl;
#else
#define debug(x)
#define debug_vec(x)
#define debug_vecvec(x)
#endif

#define rep(i,n) for(size_t i=0;i<(n);++i)
ll N;
ll min_X = INFINITY;
ll binaly_search(ll a, ll left, ll right) {
    ll middle = (left + right) / 2;
    ll now_X = pow(a,3)+pow(a,2)*middle+a*pow(middle,2)+pow(middle,3);
    if (left == right || left == middle) {
        if (now_X >= N) {
            return now_X;
        } else {
            now_X = pow(a,3)+pow(a,2)*right+a*pow(right,2)+pow(right,3);
            return now_X;
        }
    }
    if (now_X == N) {
        return now_X;
    } else if (now_X > N) {
        right = middle;
    } else {
        left = middle;
    }
    return binaly_search(a, left, right);
}
int main(){
    ll a, b;
    cin >> N;
    a = pow(10, 6);

    while(a>-1) {
        ll right = a;
        ll left = 0;
        ll i_X = binaly_search(a, left, right);
        if (i_X >= N) {
            min_X = min(min_X, i_X);
        }
        a--;
    }
    cout << min_X << endl;
    return 0;
}

File: ABC/246/d.py
import math
import sys
from typing import Union
sys.setrecursionlimit(10**9)

min_x = math.inf
N = int(input())

def calc_x(a: int, b: int) -> int:
    return a**3 + a**2 * b + a * b**2 + b**3

def binaly_search(a: int, left: int, right: int) -> Union[int, float]:
    if right < left:
        return calc_x(a, left)
    middle = (left + right) // 2
    now_x = calc_x(a, middle)
    if now_x == N:
        return now_x
    elif now_x < N:
        left = middle + 1
        match_x = binaly_search(a, left, right)
    elif now_x > N:
        right = middle - 1
        match_x = binaly_search(a, left, right)
    else:
        raise Exception('unexpected')
    return match_x

def main():
    a = 10 ** 6
    global min_x
    for i in range(a+1):
        now_x = binaly_search(i, 0, i)
        if (now_x >= N): min_x = min(min_x, now_x)
    print(min_x)
    return

if __name__ == "__main__":
    main()
